src/samples/audio/smart_amp_test.c

```diff
@@ -86,6 +142,212 @@ static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
+			 cfg_size, sizeof(struct sof_smart_amp_config));
+		return -EINVAL;
+	}
+
+	comp_dbg(dev, "smart_amp_set_config(): config size = %u", cfg_size);
+
+	memcpy_s(&sad->config, sizeof(struct sof_smart_amp_config), cfg,
+		 sizeof(struct sof_smart_amp_config));
+
+	return 0;
+}
+
+static inline int smart_amp_get_config(struct comp_dev *dev, char *data,
+				       uint32_t *data_size)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	uint32_t cfg_size;
+	int ret;
+
+	cfg_size = sizeof(struct smart_amp_data);
+
+	if (cfg_size > *data_size) {
+		comp_err(dev, "smart_amp_get_config(): wrong config size %d",
+			 *data_size);
+	}
+
+	*data_size = cfg_size;
+
+	ret = memcpy_s(data, cfg_size, &sad->config, cfg_size);
+	assert(!ret);
+
+	return 0;
+}
+
+static int smart_amp_set_large_config(struct comp_dev *dev,
+				      uint32_t param_id,
+				      bool first_block,
+				      bool last_block,
+				      uint32_t data_offset,
+				      const char *data)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "smart_amp_set_large_config()");
+
+	switch (param_id) {
+	case SMART_AMP_SET_MODEL:
+		return ipc4_comp_data_blob_set(sad->model_handler,
+					       first_block,
+					       last_block,
+					       data_offset,
+					       data);
+	case SMART_AMP_SET_CONFIG:
+		return smart_amp_set_config(dev, data, data_offset);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int smart_amp_get_large_config(struct comp_dev *dev,
+				      uint32_t param_id,
+				      bool first_block,
+				      bool last_block,
+				      uint32_t *data_offset,
+				      char *data)
+{
+	comp_dbg(dev, "smart_amp_get_large_config()");
+
+	switch (param_id) {
+	case SMART_AMP_GET_CONFIG:
+		return smart_amp_get_config(dev, data, data_offset);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int smart_amp_get_attribute(struct comp_dev *dev,
+				   uint32_t type,
+				   void *value)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "smart_amp_get_attribute()");
+
+	switch (type) {
+	case COMP_ATTR_BASE_CONFIG:
+		*(struct ipc4_base_module_cfg *)value = sad->config.base;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int smart_amp_bind(struct comp_dev *dev, void *data)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct comp_buffer __sparse_cache *buffer_c;
+	struct comp_buffer *source_buffer;
+	struct list_item *blist;
+
+	comp_dbg(dev, "smart_amp_bind()");
+
+	/* searching for feedback source buffers */
+	list_for_item(blist, &dev->bsource_list) {
+		source_buffer = container_of(blist, struct comp_buffer, sink_list);
+		buffer_c = buffer_acquire(source_buffer);
+		if (IPC4_SINK_QUEUE_ID(buffer_c->id) == SOF_SMART_AMP_FEEDBACK_QUEUE_ID) {
+			sa_lock(sad);
+			sad->feedback_buf = source_buffer;
+
+			buffer_c->stream.channels = sad->config.feedback_channels;
+			buffer_c->stream.rate = sad->config.base.audio_fmt.sampling_frequency;
+			buffer_release(buffer_c);
+			sa_unlock(sad);
```

Comment:
locks should be taken and released symmetrically. `buffer_acquire()` can potentially take a lock if the buffer is shared, so `buffer_unlock()` should come after `sa_unlock()`.  Also below you acquire the buffer inside `sa_lock()` blocks. I know, that this is just a test, so probably it cannot race, right? Still it would be better to have that consistent to not even look like we can get an AB-BA deadlock

Reply:
good catch!  thanks! refined it.

Reply:
AB-BA is still here: in this function buffer-acquisition is outside of `sa_lock()`, and in `smart_amp_trigger()` it's the other way round. Maybe you could just take the `sa_lock()` outside of the loop here?

Reply:
yes, it is a potential risk

Reply:
right, let's get it correct then. It should be ok to have `sa_lock()` outside of buffer locking since with Zephyr they're both taking mutexes and with XTOS - spinlocks

Reply:
@RanderWang can you do this in prepare() instead of bind?

Reply:
@ranj063 we can't do it in prepare for playback stream since the feedback buffer is shared by capture feedback stream, so it can be available only when the capture feedback stream is started.

Reply:
I will submit a patch later that remove bind/unbind interface and use notifier to process bind/unbind event. please check https://github.com/thesofproject/sof/pull/6751#discussion_r1046773321